### ISSUE TITLE
SCYLLA_RELEASE isn't updated when building scylla incrementally

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#8255]
I hit this when trying to bisect an OOM issue and I saw that scylla prints the wrong `--version` after I checked out and rebuilt a different version.  (At least the `--build-id` was different as it is self-contained in the binary)

Also, it is wrong to generate `SCYLLA-{RELEASE,VERSION}-FILE` under `build/`, they should sit under `build/$mode/`
See #8254

For example: (following run in dbuild)
```
$ git checkout 2d9feaace && git submodule update --init --recursive && rm -rf build && ccache -C && ./configure.py && ninja -j32 dev && build/dev/scylla --version
...
4.5.dev-0.20210310.2d9feaace

$ cat build/SCYLLA-RELEASE-FILE 
0.20210310.2d9feaace
```

```
$ git checkout 1287a5e1d && git submodule update --init --recursive && ninja -j32 build/dev/scylla && build/dev/scylla --version
HEAD is now at 1287a5e1d test: index_reader_assertions: fix misuse of trichotomic comparator in has_monotonic_positions
[1/2] NINJA build/dev/seastar/libseastar.a
ninja: Entering directory `build/dev/seastar'
ninja: no work to do.
4.5.dev-0.20210310.2d9feaace
```

`build/dev/scylla` wasn't rebuilt, despite the version change, say we can live with that...

now let's try a full build:
```
ninja -j32 dev && build/dev/scylla --version
...
4.5.dev-0.20210310.2d9feaace

$ cat build/SCYLLA-RELEASE-FILE 
0.20210310.2d9feaace
```

Still not updated :-(

So let's checkout a version that does cause scylla to be rebuilt (and updates the `tools/java` submodule)
```
$ git checkout ad563c627 && ninja -j32 build/dev/scylla && build/dev/scylla --version
...
Previous HEAD position was 1287a5e1d test: index_reader_assertions: fix misuse of trichotomic comparator in has_monotonic_positions
HEAD is now at ad563c627 Update tools/java submodule
[1/3] NINJA build/dev/seastar/libseastar.a
ninja: Entering directory `build/dev/seastar'
ninja: no work to do.
[3/3] LINK build/dev/scylla
4.5.dev-0.20210310.2d9feaace

$ cat build/SCYLLA-RELEASE-FILE 
0.20210310.2d9feaace
```

This is plain wrong.

let's try rebuilding everything again
```
$ ninja -j32 dev && build/dev/scylla --version
...
4.5.dev-0.20210310.2d9feaace
```

Only re-running configure.py updates the version files:
```
$ ./configure.py && ninja -j32 dev && build/dev/scylla --version
...
4.5.dev-0.20210310.ad563c627
```

The bottom line is that it is required to run configure.py every time before rebuild-incrementally.
Can/should this be improved so to streamline the incremental build process?

**Comment by avikivity** [comment]
It is awkward. There's a problem in that we depend on the output of git and not on the state of the filesystem, to generate the git hash. This isn't what ninja likes. Maybe there's a workaround.

Another problem is the datestamp, but it could be replaced by the datestamp of HEAD instead.

**Comment by avikivity** [comment]
There is this:

>restat
>
>if present, causes Ninja to re-stat the command’s outputs after execution of the command. Each output whose modification time the command did not change will be treated as though it had never needed to be built. This may cause the output’s reverse dependencies to be removed from the list of pending build actions.

So if we teach SCYLLA_VERSION_GEN not to touch files if they are unchanged, we could add a ninja rule to call it unconditionally. It would slow down builds a little, though.
